[processor/transform] Convert between sum and gauge in metric context

graphaelli opened the pull request:
**Description:** Allow running OTTL `convert_sum_to_gauge` and `convert_gauge_to_sum` in metric context instead of datapoint when `processor.transform.ConvertBetweenSumAndGaugeMetricContext` is enabled

closes #20773

<!--
**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>
-->

This is the result of an effort at contribfest at Kubecon NA '23.

graphaelli committed:
[processor/transform] convert_sum_to_gauge in metric context

graphaelli committed:
[processor/transform] convert_gauge_to_sum in metric context

graphaelli committed:
rename gate for using metric context for conversion between sum and gauge

graphaelli committed:
add changelog entry

graphaelli committed:
Merge branch 'main' into metric-conversion-context

graphaelli commented:
Just realized drafts don't request reviews.  Applied the same changes for`convert_gauge_to_sum`, renamed the feature gate and added a changelog entry - ready for review now.

graphaelli committed:
add missing files

graphaelli committed:
Merge branch 'main' into metric-conversion-context

fchikwekwe committed:
feat: add ability to filter and export markers (#27862)

**Description:** This add logic to filter logs based on log conditions
and sent desired logs as event markers to the honeycomb marker api.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27666

**Testing:** Unit testing for log exporter and config. Added component
testing to `otelcontribcol`.

**Documentation:** README describing component usage

Screenshot of exported markers showing up in Honeycomb
<img width="1225" alt="Screenshot 2023-11-14 at 1 27 49 PM"
src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/35741033/128d689a-cf1e-4959-9df3-6c88248a7fdb">

---------

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

a-thaler committed:
[receiver/receivercreator] Add support for k8s service discovery (#29022)

**Description:**
- Added a new watch to the k8s_observer extension for k8s services,
which can be enabled using a new flag "observe_services".
- Discovered entities are transformed into a new endpoint type
`k8s.service`.
- Adjusted the receivercreator to support the new type `k8s.service`


**Link to tracking Issue:**
[#29021](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29021)

**Testing:** Added unit tests analogue to the available tests

**Documentation:** Adjusted readme's of k8s_observer and
receivercreator. Added description of new flags and typers.

**Note:**
Current implementation is working as described in the linked ticket.
Please check the potential discussion points mentioned in the ticket:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29021#issuecomment-1801290614

---------

Co-authored-by: Antoine Toulme <antoine@toulme.name>

etiennep committed:
feat: add IsDouble function (#29076)

**Description:** Adds new a new `IsDouble` function to facilitate type
checking. Most useful when checking the type of a body to determine if
it needs to be parsed or not.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27895

**Testing:** Added unit test

**Documentation:** Updated the func readme.

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

bryan-aguilar committed:
[exporter/awsemf] Improve summary metric type NaN checks (#28894)

**Description:** I have observed some behavior on a personal collector
deployment where the EMF Exporter is still returning errors for `NaN`
json marshalling. This was in a prometheus -> emf exporter metrics
pipeline.

I could not find the specific NaN value in the metrics when
troubleshooting the error. I curled the `/metrics` endpoint and also
tried using the logging exporter to try to get more information. I could
not find where the NaN value was coming from so I took another look into
the unit tests and found some possible code paths in which NaNs could
slip though.

**Link to tracking Issue:** Original issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26267

**Testing:** Added more unit tests. The summary unit tests got a slight
refactor for two reasons. So I could get ride of the unnecessary
typecasting and so that we could more easily test out different
combinations of quantile values.

I have also added a few more histogram unit tests to just verify that
all combinations of NaN values are being checked on their own.

rajkumar-rangaraj committed:
[exporter/azuremonitor] Enhance AAD Documentation. (#29234)

**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
* Update AAD documentation to use connection string instead of
instrumentation key. Follow up to #28854
* Modified the ingestion version from 2.0 to 2.1

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were
added.>

Existing tests.

Output from manual run

``` json
--------- Transmitting 30 items ---------       {"kind": "exporter", "data_type": "logs", "name": "azuremonitor"}
2023-11-13T10:50:23.886-0800    debug   azuremonitorexporter@v0.88.0/factory.go:139     Telemetry transmitted in 378.439395ms   {"kind": "exporter", "data_type": "logs", "name": "azuremonitor"}
2023-11-13T10:50:23.886-0800    debug   azuremonitorexporter@v0.88.0/factory.go:139     Response: 200   {"kind": "exporter", "data_type": "logs", "name": "azuremonitor"}
2023-11-13T10:50:23.886-0800    debug   azuremonitorexporter@v0.88.0/factory.go:139     Items accepted/received: 30/30 {"kind": "exporter", "data_type": "logs", "name": "azuremonitor"}
```

**Documentation:** <Describe the documentation added.>
* Updated Authentication.md

TylerHelmuth committed:
[exporter/honeycombmarker] Fix default url and dataset_slug (#29309)

**Description:** 
Fixes an issue with an incorrect default url. Also fixes issue where
dataset slug was required.

**Link to tracking Issue:** <Issue number if applicable>
Related to
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27666

**Testing:** <Describe what testing was performed and which tests were
added.>
Added new tests and tested manually.

**Documentation:** <Describe the documentation added.>
 Updated up README

fchikwekwe committed:
feat: change honeycombmarkerexporter to alpha (#29272)

**Description:** Update Honeycomb Marker Exporter to alpha status

**Link to tracking Issue:** #27666 

**Testing:** 

**Documentation:**

---------

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

TylerHelmuth committed:
[exporter/honeycombmarker] Add generated README header (#29311)

crobert-1 committed:
[chore][pkg/translator/azure] Add missing code owner @cparkins (#28651)

**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This fixes inconsistency introduced with the creation of this package.
In #25096 @cparkins was added as a code owner in the
[metadata.yaml](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/azure/metadata.yaml)
but not the top level `CODEOWNERS` file.

Co-authored-by: Alex Boten <aboten@lightstep.com>

evan-bradley commented on CHANGELOG.md:
Thanks for adding a changelog message. This file is autogenerated, use `make chlog-new` to generate a new changelog entry file, and edit that file to create the entry.

graphaelli replied:
Thanks, done.

evan-bradley commented on processor/transformprocessor/internal/metrics/functions_test.go:
```diff
@@ -31,6 +31,8 @@ func Test_DataPointFunctions(t *testing.T) {
 
 func Test_MetricFunctions(t *testing.T) {
```
I think you'll need to toggle the feature gate in this test.

graphaelli replied:
thanks, done

evan-bradley reviewed (commented):
Thanks for contributing this! Everything mostly looks good to me.

TylerHelmuth committed:
[exporter/honeycombmarker] Update distributions (#29312)

jacobmarble committed:
[exporter/influxdb] fix panic on init (#29296)

When InfluxDB v1 compatibility is enabled AND username&password are set,
the exporter panics. Not any more!

Fixes #27084

**Testing:** I've added one regression test.

TylerHelmuth committed:
[chore] downgrade golangci-lint (#29316)

https://github.com/open-telemetry/opentelemetry-collector/issues/8939

Co-authored-by: Alex Boten <aboten@lightstep.com>

Commit:
[chore] fix issuegenerator path (#29318)

workflows have been failing and then trying to use `issuegenerator` to
create issues, but the path for the tool was incorrect. see
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6895702499/job/18761957296
as an example

Signed-off-by: Alex Boten <aboten@lightstep.com>

TylerHelmuth committed:
[receiver/kafkametrics] skip flaky test (#29317)

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26293

JonathanWamsley committed:
[receiver/mongodbatlasreceiver] add metrics project config (#28866)

**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This feature adds a Project Config for the metrics to filter by Project
name and or clusters.

**Link to tracking Issue:** <Issue number if applicable>
#28865 

**Testing:** <Describe what testing was performed and which tests were
added.>
- Added test for cluster filtering
- Tested project name alone, project name with IncludeClusters and
project name with ExcludeClusters on a live environment with success.

**Documentation:** <Describe the documentation added.>
Added optional project config fields to README

---------

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

atoulme committed:
[chore][checkapi] Do not enforce that all packages must have exactly one function (#28886)

If no functions are exposed, exit with no error.

This change allows to remove `extension/encoding` from the allowlist.

graphaelli committed:
adjust test setup for metric context

graphaelli committed:
toggle feature gate during test

per review feedback

graphaelli committed:
Replace manual changelog entry with chloggen

graphaelli committed:
Merge branch 'main' into metric-conversion-context

graphaelli commented:
not sure what happened there, thought I just merged main back in.

graphaelli committed:
remove unused struct

graphaelli committed:
crosslink

graphaelli commented:
@evan-bradley thanks for the first pass.  Addressed all comments and ready for a review.

graphaelli committed:
gotidy

evan-bradley commented:
@graphaelli Sorry, looks like the go.mod changed. Could you rebase one more time?

graphaelli committed:
Merge branch 'main' into metric-conversion-context

graphaelli committed:
Merge branch 'main' into metric-conversion-context
